packages/core/src/modules/connections/handlers/ConnectionResponseHandler.ts

```diff
@@ -65,8 +66,26 @@ export class ConnectionResponseHandler implements Handler {
     }
 
     messageContext.connection = connectionRecord
-    // The presence of outOfBandRecord is not mandatory when the old connection invitation is used
-    const connection = await this.connectionService.processResponse(messageContext, outOfBandRecord)
+
+    const invitationRecipientKeys = outOfBandRecord
+      .getTags()
+      .recipientKeyFingerprints.map((fingerprint) => Key.fromFingerprint(fingerprint))
```

Comment:
I think we should store the invitation recipient keys with public did keys from the start rather than in the .

My proposal would be to move the `recipientKeyFingerprints` from the default tags (generated on save) to custom tags that we must set ourselves. When processing the invitation we can set all recipient keys. The advantage is that we only need to do it once for both connection and didexchagne protocol, and we also don't need this or functionality (we now only add the did recipient keys if we don't have any inline didcomm service keys).

Reply:
Yes, I get the point and fully agree that it would be better to store the recipient keys at the start of the connection process. However I'm not sure where exactly to do that. To get the recipient keys, we need to run the did resolution for services from the ledger (which is quite costly operation).
I thought of 2 options:
1. Either to do it right when creating the OOB record in `OutOfBandModule.receiveInvitation`. Which would mean to run an additional resolution. Also I'm not sure what to put there in case when we are creating the invitation. `OutOfBandModule.createInvitation`
2. Store the recipientKeys into the record once we have already resolved them (when sending the connection request message in `MessageSender.retrieveServicesByConnection`): https://github.com/hyperledger/aries-framework-javascript/blob/49f8a1b9f9e1c9223755f39f5445ff2fc89e8627/packages/core/src/agent/MessageSender.ts#L363)

I'm also not familiar with the concept of the custom/default tags to be honest. Currently the recipientKeyFingerprints are retrieved from the OOB invitation object: https://github.com/hyperledger/aries-framework-javascript/blob/49f8a1b9f9e1c9223755f39f5445ff2fc89e8627/packages/core/src/modules/oob/repository/OutOfBandRecord.ts#L69 Does it mean they are not saved as tags? Or am I missing something?

Reply:
> Which would mean to run an additional resolution

Where would we need the additional resolution for?

> I'm also not familiar with the concept of the custom/default tags to be honest. Currently the recipientKeyFingerprints are retrieved from the OOB invitation object:

We have two types of tags in AFJ; default and custom. Default tags are set in the `getTags` method on a record. This can contain derived properties (such as the recipientKeysFingerprints`, but can't be async. This worked with inline services, but if we want to get the recipientKeyFingerprints for the public dids we need to do the async resolving. What we can do is remove it from the default tags (so we're not setting it anymore in the getTags method), and rather set it manaually using `record.setTag('recipientKeyFingerprints', theKeys)`.

You would have to create a new type `CustomOutOfBandRecordTags` type, and add it here: https://github.com/hyperledger/aries-framework-javascript/blob/49f8a1b9f9e1c9223755f39f5445ff2fc89e8627/packages/core/src/modules/oob/repository/OutOfBandRecord.ts#L26-L33

```
type DefaultOutOfBandRecordTags = {
  role: OutOfBandRole
  state: OutOfBandState
  invitationId: string
}

type CustomOutOfBandRecordTags = {
  recipientKeyFingerprints?: string[]
}

export class OutOfBandRecord extends BaseRecord<DefaultOutOfBandRecordTags, CustomOutOfBandRecordTags> {
```

Reply:
Great, thank you for the explanation of the tags.

> > Which would mean to run an additional resolution
>
> Where would we need the additional resolution for?

I meant if we now additionally run a resolution right in the beggining when creating the oob record, that would be the additional resolution. I'll try with the approach n°2, to set the recipientKeyFingerprints after sending the credential request message (for the case of public did invite), otherwise set it immediately when creating the oob record from the invitation's inline services.